daemon/restart.go

```diff
@@ -39,8 +39,10 @@ func (daemon *Daemon) containerRestart(container *container.Container, seconds i
 		return err
 	}
 
-	if err := daemon.containerStart(container); err != nil {
-		return err
+	if !container.HostConfig.AutoRemove {
```

Comment:
I totally agree with you about the behavior, but I think remove this if clause won't make it work. Maybe I need to do some trick, for example, set `--rm=false` before restart, and restore it after restart. I'll make some try.